good design of alignment traits

MattesMrzik opened the pull request:
- implemented traits Alignment and AncestralAlignment, closes #25
- ASR trait that takes Alignment and outputs AncestralAlignment, closes #51 
- Fixes #62

MattesMrzik committed:
good design of alignment traits?

- AlignmentTrait and AncestralAlignmentTrait
- Asr trait that takes AlignmentTrait and outputs AncestralAlignmentTrait

junniest commented on phylo/src/substitution_models/tests.rs:
```diff
@@ -452,7 +452,7 @@ fn designation() {
 
 #[cfg(test)]
-fn setup_simple_phylo_info(blen_i: f64, blen_j: f64) -> PhyloInfo {
+fn setup_simple_phylo_info(blen_i: f64, blen_j: f64) -> PhyloInfo<impl AlignmentTrait> {
```
Again, in this file it should be `PhyloInfo<Alignment>`, not `PhyloInfo<impl AlignmentTrait>`, since the test actually only works with Alignment, not with anything that has that trait implemented.

junniest commented on phylo/src/substitution_models/mod.rs:
```diff
@@ -114,17 +115,17 @@ impl<Q: QMatrix> EvoModel for SubstModel<Q> {
 }
 
-pub struct SubstitutionCostBuilder<Q: QMatrix> {
+pub struct SubstitutionCostBuilder<Q: QMatrix, M: AlignmentTrait> {
```
Again, I think I'd prefer `A: Alignment` to `M: AlignmentTrait` purely because `M` is normally used for the model.

junniest commented on phylo/src/pip_model/tests.rs:
```diff
@@ -274,7 +274,7 @@ fn pip_p_example_matrix() {
 
 #[cfg(test)]
-fn setup_example_phylo_info() -> PhyloInfo {
+fn setup_example_phylo_info() -> PhyloInfo<impl AlignmentTrait> {
```
I think that here and in other places in this file it should be PhyloInfo<Alignment>, not PhyloInfo<impl AlignmentTrait>, since the test actually only works with Alignment, not with anything that has that trait implemented.

junniest commented on phylo/src/pip_model/mod.rs:
```diff
@@ -177,14 +177,15 @@ pub struct PIPModelInfo<Q: QMatrix> {
 
 impl<Q: QMatrix> PIPModelInfo<Q> {
-    pub fn new(info: &PhyloInfo, model: &PIPModel<Q>) -> Result<Self> {
+    pub fn new<M: AlignmentTrait>(info: &PhyloInfo<M>, model: &PIPModel<Q>) -> Result<Self> {
```
I'd prefer `A: Alignment` to `M: AlignmentTrait` purely because M is normally used for the model.

junniest commented on phylo/src/phylo_info/tests.rs:
```diff
@@ -14,7 +14,7 @@ use crate::{frequencies, record_wo_desc as record, tree};
 #[cfg(test)]
 fn downcast_error<T: Display + Debug + Send + Sync + 'static>(
-    result: &Result<PhyloInfo, anyhow::Error>,
+    result: &Result<PhyloInfo<impl AlignmentTrait + Debug>, anyhow::Error>,
```
Hm, is the Debug requirement necessary? Maybe it should be enforced in `AlignmentTrait` instead if so (`pub trait AlignmentTrait: Debug`)?

MattesMrzik replied:
Currently, this fn is only used with the struct MSA. So I replaced it with this.

junniest commented on phylo/src/phylo_info/phyloinfo_builder.rs:
````diff
@@ -141,7 +141,7 @@ impl PhyloInfoBuilder {
     /// assert_eq!(info.tree.len(), 7);
     /// ```
-    pub fn build(self) -> Result<PhyloInfo> {
+    pub fn build(self) -> Result<PhyloInfo<impl AlignmentTrait>> {
````
At the moment the function will create an instance of `PhyloInfo<Alignment>` no matter what is being passed here, which might be misleading. This means that we need to maybe put creation of Alignments (`Alignment::from_aligned(sequences, &tree)?`) into the trait as well. This creates some other problems unfortunately, but in any case at the moment this particular parametrisation makes no difference and I'd rather restrict it to `PhyloInfo<Alignment>` to avoid confusion.

MattesMrzik replied:
Yep. If I implement the TKF stuff, then I need a PhyloInfo that contains an AncestralAlignment. I was thinking about adding a new field `asr: Option<Box dyn AncestralSequenceReconstruction>` to `PhyloInfo`. Then, when the build method is called, a AncestralAlignment is build and returned as a field of the `PhyloInfo<impl AlignmentTrait>`.

junniest replied:
Hm, you meant adding the field to `PhyloInfoBuilder`, right? Yeah, probably makes sense, or at least until I come up with a way to refactor this to be better. And pure ASR will happen only once on initialisation? Then the other ancestral sequences will be estimated on the go and the ASR won't be needed again?

MattesMrzik replied:
Yes, I meant `PhyloInfoBuilder`. I have updated the comment. Yes, this `ASR` will only happen once.

junniest commented on phylo/src/phylo_info/mod.rs:
```diff
@@ -16,15 +16,15 @@ pub use phyloinfo_builder::*;
 /// * Add support for unaligned sequences.
 #[derive(Debug, Clone)]
-pub struct PhyloInfo {
+pub struct PhyloInfo<M: AlignmentTrait> {
```
I'd prefer `A: Alignment` here, since `M` is used for models

junniest commented on phylo/src/phylo_info/mod.rs:
```diff
@@ -16,15 +16,15 @@ pub use phyloinfo_builder::*;
-impl PhyloInfo {
-    /// Compiles a represenataion of the alignment in a vector of fasta records.
+impl<M: AlignmentTrait> PhyloInfo<M> {
+    /// Compiles a representation of the alignment in a vector of fasta records.
```
Lol thank you for fixing that

junniest commented on phylo/src/parsimony_indel_sites/tests.rs:
```diff
@@ -0,0 +1,65 @@
+use crate::alignment::{Alignment, AlignmentTrait, AncestralAlignmentTrait, Sequences};
+use crate::asr::Asr;
```
yeah, abbreviations that are not all caps look weird to me, and it's a niche term, so I'd rather go with `AncestralReconstruction`.

junniest commented on phylo/src/parsimony_indel_sites/tests.rs:
```diff
@@ -0,0 +1,65 @@
+use crate::{align, record, tree};
+
+#[cfg(test)]
+fn test_tree() -> Tree {
```
Nitpick: You likely don't need this as a separate function, it can be embedded in the test itself, then it's a bit easier to read imo.

junniest reviewed (commented):
You're definitely on the right track, awesome work!

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -31,12 +31,34 @@ impl PairwiseAlignment {
+    fn leaf_map(&self, node: &NodeIdx) -> &Mapping;
+    fn compile_subroot(&self, subroot: Option<&NodeIdx>, tree: &Tree) -> Result<Sequences>;
+    fn leaf_encoding(&self) -> &HashMap<String, DMatrix<f64>>;
+    fn int_align_map(&self) -> &InternalAlignments;
```
Yes, I am happy with `internal_alignments`. I would Also rename `leaf_map` to `leaf_maps` since it contains the `Mapping`s for all leafs. But if use bitmaps instead of mappings the name might have to change again.

junniest commented:
> * AlignmentTrait and AncestralAlignmentTrait
> * Asr trait that takes AlignmentTrait and outputs AncestralAlignmentTrait
> 
> * `fmt` of Alignment prints unaligned sequences, which is unintuitive for me. I added idx_to_id vec in AncestralAlignment to print aligned seqs in fmt.

Oh absolutely, you are correct, it's counterintuitive because I didn't think it through and just printed whatever was easiest. `Display` should certainly print the aligned sequences.

> * the remarks you made in PR [Feature/Implements ancestral alignment #48](https://github.com/acg-team/rust-phylo/pull/48) about names of bools in the elongate_maps_and_seqs() will be addressed in the upcoming commits
> * Question:
>   
>   * For me, the old name of `node_map` feels unintuitive. I prefer naming vars after their content and not type. Since the node_map is HashMap<NodeIdx, PairwiseAlignment> I would call it `internal_alignments`. (Currently, I somewhat tried to keep the naming conventions I found in existing code.)

Again, fully agree. I think I was just running out of brainpower to make up meaningful names and resorted to what you see. I appreciate you sticking to what was there originally, but please, feel free to give members more meaningful names.

> * Question:
>   
>   * Since `PhyloInfo` has generic field of type AlignmentTrait, structs that have PhyloInfo as field now also must implement the generic type AlignmentTrait. Would it make sense to refactor to use field `msa: Box<dyn AlignmentTrait>` as field for PhyloInfo instead?

Hm. For some reason I really don't want to use `Box`es because it makes me think that it will slow down the code, but I am not entirely sure that it is valid. If I understand correctly, a `Box` would mean that the methods are dispatched dynamically (the type is determined in runtime which adds overhead), whereas using a generic will mean that the dispatching will be done in compile time. It may lead to negligible change in performance though.

I am not sure what is best at the moment. Let's still think about it.

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -16,8 +16,8 @@ pub use aligner::*;
 
 pub type Position = Option<usize>;
 pub type Mapping = Vec<Option<usize>>;
```
I opened Issue #54. There, we can discuss and implement this.

MattesMrzik committed:
Merge branch 'develop' into #25-Acestral-Alignment-as-trait

MattesMrzik committed:
Merge upstream, generics for alignment+aligner+asr

MattesMrzik committed:
bug fix, adjusted for removed leaf_encoding

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -31,12 +31,34 @@ impl PairwiseAlignment {
 
 #[derive(Debug, Clone)]
 pub struct Alignment {
+    // TODO: i could remove the pub crate here if i would write a ctor and getters, and maybe also adjust the compile_leaf_map()
```
Lets talk in person about this next time:)

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -281,6 +209,227 @@ impl Alignment {
+        }
+    }
+
+    /// the seqs should contain a seq for every node in the tree
```
The matching of ids of the tree nodes and ids of the seqs is also a requirement in the Alignment::from_aligned but there it is also not checked if they match. Should the check also be added there? (currently it panics i think)

MattesMrzik commented on phylo/src/likelihood/tests.rs:
```diff
@@ -70,11 +70,11 @@ fn protein_search_costs_equal() {
     freqs: &[f64],
     params: &[f64],
-) -> PIPCost<Q> {
+) -> PIPCost<Q, Alignment> {
```
In the tests we need to a concrete type, so until we have different implementations of the Alignment trait we dont need to make this method generic. Does this make sense?

MattesMrzik committed:
compile subroot for MASA, minor changes

- minor changes are implemented remarks from julija

MattesMrzik committed:
phylo builder with ancestors and tests

MattesMrzik committed:
clean up before upsteam merge

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into #25-Acestral-Alignment-as-trait

MattesMrzik committed:
Minor cleanup after merge

MattesMrzik committed:
fix doc test

MattesMrzik committed:
more tests

MattesMrzik committed:
more tests

junniest commented on phylo/src/parsimony_indel_points/mod.rs:
```diff
@@ -0,0 +1,111 @@
+pub struct ParsimonyIndelPoints {}
+
+impl<A: Alignment, AA: AncestralAlignment> AncestralSequenceReconstruction<A, AA>
+    for ParsimonyIndelPoints
```
I said that above somewhere, but I'll mention it here too, I don't particularly like the `ParsimonyIndelPoints` name because to me it implies that this only reconstructs indel points and not the characters.

junniest commented on phylo/src/parsimony_indel_points/mod.rs:
```diff
@@ -0,0 +1,111 @@
+    //       see https://github.com/acg-team/rust-phylo/pull/48/files#r2046683347
+    for site in 0..alignment.len() {
+        // upward pass
+        let has_char = get_has_char(alignment, tree, site);
```
this can be called `has_char`, since `get` is not particularly used in Rust.

junniest replied:
I wonder if this is something we need to recompute with a pass through the tree or could be computed on the go as well. The issue is that going through sites then nodes is always going to be slower than going through nodes then sites because the tree is not a continuous structure and will need a lot of memory access.

junniest replied:
Moreover, `get_has_char` accesses the sequences as well, which increases complexity by a lot.

MattesMrzik replied:
Will be addressed in #63

junniest commented on phylo/src/tree/mod.rs:
```diff
@@ -68,6 +68,7 @@ pub struct Tree {
     preorder: Vec<NodeIdx>,
     leaf_ids: Vec<String>,
     pub complete: bool,
+    /// number of leaves
```
Nitpick: Maybe change this to `number of leaves/sequences` just to remind people?

junniest commented on phylo/src/phylo_info/tests.rs:
```diff
@@ -366,3 +365,150 @@ fn force_protein_alphabet() {
+
+    // assert
+    let error_msg = res_info.unwrap_err().to_string();
+    assert!(error_msg.contains("Building an ancestral alignment from unaligned sequences (including ancestral_sequencess) is not supported"));
```
Nitpick: typo `sequencess` -> `sequences`

junniest replied:
Haha, so reconstruction is implemented now but regular input is not? (: fair enough!

junniest commented on phylo/src/phylo_info/tests.rs:
```diff
@@ -366,3 +365,150 @@ fn force_protein_alphabet() {
+    // arrange
+    let fldr = PathBuf::from("./data");
+    let builder = PIB::with_attrs(
+        fldr.join("sequences_DNA1_missmatched_id.fasta"),
```
Nitpick: typo `missmatched` -> `mismatched`

junniest commented on phylo/src/phylo_info/phyloinfo_builder.rs:
```diff
@@ -1,25 +1,30 @@
     tree_file: Option<PathBuf>,
-    alignment_builder: Option<Box<dyn Aligner>>,
+    // since alignment and asr is only done once, we can use dynamic dispatch
+    // but since we access the alignment on a regular basis (or do we actually? Since we instead use the encoding)
```
Good q, let's discuss at some point.

junniest commented on phylo/src/parsimony_indel_points/tests.rs:
```diff
@@ -0,0 +1,64 @@
+    assert_eq!(leaf_seqs.s, true_leaf_seqs.s);
+    let ancestral_seqs = ancestral_msa.ancestral_seqs();
+    let true_ancestral_seqs =
+        aligned_seqs_with_ancestors_subset(&["I5", "I9", "I7", "I8"]).into_gapless();
```
Shouldn't the gapped ones also match?

MattesMrzik replied:
add comment to say that maps already got checked

junniest reviewed (commented):
Good stuff! I think there are still some minor design things we should discuss.

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -30,27 +32,209 @@ impl PairwiseAlignment {
+    fn len(&self) -> usize;
+    fn seq_count(&self) -> usize;
+    fn leaf_map(&self, node: &NodeIdx) -> &Mapping;
+    fn compile_subroot(&self, subroot: Option<&NodeIdx>, tree: &Tree) -> Result<Sequences>;
```
When implementing my suggested changes, I try to not fiddle too much with preexisting code. Just to keep the scope of the PR reasonable. Since the `PhyloInfo` calls `compile_subroot` on its field msa, and this msa is generic, it needs to be part of the trait. 
So I agree, this should go into another PR, perhaps also into #54.

MattesMrzik replied:
I will move this to the `phylo`, where we have access to the correct tree

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
+
+impl Display for MASA {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let both_maps = self.leaf_maps.iter().chain(self.ancestral_maps.iter());
```
Added this to #62

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
+        &self.internal_alignments
+    }
+
+    // TODO: or do i want to include the ancestral seqs here?
```
On the one hand this makes sense, however, the idea behind implementing the `Alignment` trait also for `MASA` is that we can use MASA also wherever we just want to have a (leaf) Alignment. In those places, we would not expect ancestral sequences.

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
+        };
+        let leaf_map = alignment.compile_leaf_map(&tree.root, tree).unwrap();
+        alignment.leaf_maps = leaf_map;
+        let asr = ParsimonyIndelPoints {};
```
> Would we like to be able to use different ASR methods here?

We just need an implementation to please the trait. If we want to have an ancestral alignment with our ASR of choice we would call MSA::from_aligned, pass this to the desired ASR approach, and get our MASA this way.

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
-            }
+    /// Assumes that nodes of the tree have unique ids and that the leaf ids match the ids of the
+    /// sequences.
+    fn from_aligned(sequences: Sequences, tree: &Tree) -> Result<Self> {
```
or unchecked_...

MattesMrzik committed:
renamed parsimony wildcard reconstruction

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
+        &self.ancestral_seqs
+    }
+
+    /// Assumes that nodes of the tree have unique ids and that they match the ids of the
```
checks are done now

MattesMrzik committed:
default checks are defined in the trait
- renamed parsimony_indel_points to parsimony_presence_absence
- moved the compile alignment methods to phylo_info
- in the aligner the interanl alignments are not used directly used in
  the new alignment but instead to get aligned sequences, then the
  from_aligned is called (which rebuilds the internal_alignments)

MattesMrzik committed:
Merge remote-tracking branch 'upstream' into #25-Acestral-Alignment-as-trait

MattesMrzik committed:
fixed small bugs, added tests

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into #25-Acestral-Alignment-as-trait

MattesMrzik committed:
fix doc test

junniest commented on phylo/src/alignment/aligner.rs:
```diff
@@ -1,7 +1,19 @@
+use crate::phylo_info::validate_taxa_ids;
+
+pub trait Aligner<A: Alignment> {
+    fn align(&self, seqs: &Sequences, tree: &Tree) -> Result<A> {
```
I think that this is ok for now, but if we provide this as a generic trait implementation we might want to make this also work for potential alternative implementations of the tree and sequences. That would mean that `seqs.ids_are_unique()` and `validate_taxa_ids(seqs, tree)` would be somehow generic too.

I think that I could refactor the tree to be a trait rather than a struct, a developer friend suggested that allowing users of the crate to use their own types for nodes and trees is important. I think I could also refactor the Sequences object to be iterable and accept an iterator of Records here...

Although, this might be too much generalisation. I will think about it, but I think this is good for now!

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -30,27 +34,100 @@ impl PairwiseAlignment {
     }
 }
+#[allow(clippy::len_without_is_empty)]
+pub trait Alignment: Display + Clone + Debug {
```
I think we need docstrings for the traits as well. Not sure if we need examples here, and I can't seem to find any relevant guidelines in Rust. Let me know if you find any!

MattesMrzik replied:
I added some

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -60,19 +137,23 @@ impl Alignment {
     ///     record!("C2", Some("C2 sequence"), b"AA--"),
     /// ], dna_alphabet());
-    /// let msa = Alignment::from_aligned(seqs, &tree).unwrap();
+    /// let msa = MSA::from_aligned(seqs, &tree).unwrap();
```
One very important guideline in Rust is that the doctests should not use `.unwrap` [link](https://rust-lang.github.io/api-guidelines/documentation.html#examples-use--not-try-not-unwrap-c-question-mark). The reason is that users copy examples directly (lol) and should be aware that there's an unwrap going on. We should use a question mark for doctests, which means you need to wrap the code in a hidden main function that returns `Result`

In doctest terms it looks like this:
```
# fn main() -> std::result::Result<(), anyhow::Error> {
rest of the code; 
# Ok(()) }
```

There's an example in this file: [phylo/src/optimisers/topo_optimiser.rs](https://github.com/acg-team/rust-phylo/blob/966f39a804a9addf6f4b60ef4b56ed5d0abb6008/phylo/src/optimisers/topo_optimiser.rs).

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,52 +188,64 @@ impl Alignment {
+        self.leaf_maps.len()
+    }
+
+    fn leaf_map(&self, node: &NodeIdx) -> &Mapping {
```
I think this is fine for now, but we may want to think how to make these generic so they don't rely on our defined `Mapping`, `SeqMaps`, etc. But, again, this might be too generic and unimplementable because of that.

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -165,114 +258,145 @@ impl Alignment {
+        &self.leaf_seqs
+    }
+    #[allow(clippy::len_without_is_empty)]
+    fn len(&self) -> usize {
```
nitpick: newline before (lol, sorry)

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
+                    leaf_records.push(record.clone());
+                }
+            };
+            idx_to_id[usize::from(node_idx)] = record.id().to_string();
```
Or, potentially, a map from IDs to both tree node indices and sequence indices in `PhyloInfo`? So that it defines a definitive mapping for both objects once they are constructed?

MattesMrzik replied:
But if we only have this in `PhyloInfo` then from within the alignment we dont know which `leaf_mapping` corresponds to what `seq_id`, so we would be unable to `fmt` the Alignment with aligned seqs. But we can think about whether such a mapping would be nice to have in `phyloInfo`.

MattesMrzik replied:
See #97

junniest commented on phylo/src/alignment/sequences.rs:
```diff
@@ -155,13 +153,24 @@ impl Sequences {
         self.s = new_seqs.collect();
     }
+
+    pub(crate) fn ids_are_unique(&self) -> Result<()> {
```
Yeah, I wonder if it's a good idea to use a `pub(crate)` method in a generic trait implementation. Probably isn't, so maybe until we refactor it should be public? I would also suggest changing the interface to return a bool rather than `Result<()>`, since the return type is irrelevant in this case.

MattesMrzik replied:
I agree with the first part.
However, for the second, i find it quite elegant to return a result. Since if ids are not unique, we usually want to bail. So, by returning Result<()>, we dont have to have the same bail msg at multiple locations in the code.

junniest replied:
Fair enough, for some reason I thought that `Result<()>` isn't idiomatic, but I was wrong (: good job!

junniest commented on phylo/src/alignment/tests.rs:
```diff
@@ -1,11 +1,11 @@
-use crate::alignment::{
-    Alignment, InternalMapping, LeafMapping, PairwiseAlignment as PA, Sequences,
-};
+use crate::alignment::{Alignment, AncestralAlignment, MASA, MSA};
```
I still haven't figured out if style-wise it's best to keep all use statements from a single submodule grouped together or try to balance them out like this. Potentially it's better to have a single line with all things from the submodule, because then the formatter will sort them, but I am happy to hear your thoughts.

MattesMrzik replied:
makes sence:)

MattesMrzik commented on phylo/src/parsimony/mod.rs:
```diff
@@ -101,14 +102,24 @@ impl<'a, PS: ParsimonyScoring + Clone> ParsimonyAligner<PS> {
-        let leaf_map = alignment.compile_leaf_map(&tree.root, tree).unwrap();
-        alignment.leaf_map = leaf_map;
-        Ok((alignment, scores))
+        // TODO: to avoid having a fn new(tree, seqs, internal_alignments, leaf_maps) for the Alignment trait,
+        // we instead get the aligned Sequences and then create the alignment from it. This discards the internal
+        // alignments and rebuilds them when calling `from_aligned`, which might not be ideal.
+        let leaf_maps = PhyloInfo::<A>::compile_leaf_map(&tree.root, &alignments, seqs, tree);
+        let aligned_seqs = Sequences::with_alphabet(
+            leaf_maps
+                .iter()
+                .map(|(idx, map)| {
+                    let rec = seqs.record_by_id(tree.node_id(idx));
+                    let aligned_seq = aligned_seq!(map, rec.seq());
+                    Record::with_attrs(rec.id(), rec.desc(), &aligned_seq)
+                })
+                .collect(),
+            *seqs.alphabet(),
+        );
+        let alignment = A::from_aligned_unchecked(aligned_seqs, tree);
+
+        (alignment, scores)
```
what about this?

junniest replied:
Hm. Looks better, I think, but we could also return the AncestralAlignment here? That should not make a difference downstream since AncestralAlignment implements Alignment, and the internal alignment can be preserved.

Also, I should really refactor this function, it's so ridiculously long now. But I think I'll get to it when I work on the other alignment stuff, hopefully soon.

MattesMrzik replied:
Mmh, I dont think that (with the current state of code) we can preserve the internal alignments, even with Ancestral alignment.

junniest replied:
Gotcha, let's leave it as is for now.

MattesMrzik committed:
added docs to Aligner, Alignment, Asr, ParsimonyPresenceAbsence

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into #25-Acestral-Alignment-as-trait

MattesMrzik committed:
added back aligned_seq macro, fixed merge bugs
- including doc tests

MattesMrzik committed:
removed unwrap from doc tests

MattesMrzik committed:
refine docs

MattesMrzik committed:
Merge remote-tracking branch 'upstream' into #25-Acestral-Alignment-as-trait

MattesMrzik committed:
more coverage

MattesMrzik committed:
more cleanup and tests

junniest commented on phylo/src/alignment/aligner.rs:
```diff
@@ -1,19 +1,25 @@
     fn align(&self, seqs: &Sequences, tree: &Tree) -> Result<A> {
         if seqs.aligned {
-            bail!("Sequences must not be aligned.");
+            bail!("Sequences must not be aligned");
```
Actually, I wonder if this should be a `bail!` scenario, or rather a `warn!("Sequences are already aligned, realigning")` and keep going scenario? I can open an issue for this to fix later, but we probably should have an `unalign` method and then align the sequences with our own approaches.

MattesMrzik replied:
Sounds good. Will open a new issue for that

MattesMrzik replied:
See #98

MattesMrzik committed:
Merge branch 'develop' into #25-Acestral-Alignment-as-trait

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -17,11 +17,17 @@ pub use sequences::*;
-pub type Mapping = Vec<Option<usize>>;
+/// Represents aligned positions of a sequence.
+/// E.g. The `Mapping` for the sequence `A--T-` is `[Some(0), None, None, Some(1), None]`.
+pub type Mapping = Vec<Position>;
```
I've been thinking about this for a while and I am wondering again if using a `FixedBitSet` would make this mapping faster? here we have to go through each position and check whether the element is `Some(x)` or `None`, but a `FixedBitSet` has a `.ones()` method, which returns an iterator over all set indices. Then something like `A--T-` will simply be represented as a `FixedBitSet` that is `10010` and `.ones()` will return [0, 3].
On the other hand, if we keep the mapping as is, we can potentially represent alignments where positions can be moved around (e.g. something weird like `[Some(3), Some(1), Some(2), Some(0)]`). Maybe it could be useful in the future, but I am not convincing myself at the moment :D
If you think we might be better off switching, I'll make an issue and then try to benchmark it to see if it makes a difference for performance or not.

MattesMrzik replied:
Agreed, see #54.

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -34,41 +40,75 @@ impl PairwiseAlignment {
 pub trait AncestralAlignment: Alignment {
     fn ancestral_seqs(&self) -> &Sequences;
     fn ancestral_map(&self, node_idx: &NodeIdx) -> &Mapping;
+    fn ancestral_maps(&self) -> &SeqMaps;
```
Do we need both methods?

MattesMrzik replied:
Not yet. But might be useful later. But can remove it for now.

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -34,41 +40,75 @@ impl PairwiseAlignment {
         all_seqs.remove_gap_cols();
         Ok(Self::from_aligned_with_ancestral_unchecked(all_seqs, tree))
     }
+    /// Constructs an ancestral alignment instance from aligned sequences and a phylogenetic tree. Is called
```
Hm. I guess we cannot really enforce in the trait that this is only called by `Self::from_aligned_with_ancestral`. While your default implementation does that, it can still be overridden. Maybe it's worth changing the doc to say "Is called by the default implementation of by [`Self::from_aligned_with_ancestral`]" and in that method state that this is the expected behaviour and people should think twice before overriding the default implementation (and copying code to end up with two versions of aligning code)...

MattesMrzik replied:
sounds good, will add a comment

junniest commented on phylo/src/parsimony_presence_absence/mod.rs:
```diff
@@ -1,13 +1,25 @@
+/// Reconstructs ancestral wild card sequences, i.e. instead of the usual reconstruction that
+/// returns sequences with characters of the alphabet, this just returns the presence or absence of
+/// any character at a site. These are represented in an [`AncestralAlignment`] as sequences of
+/// [AMB_CHAR](`crate::alphabets::AMB_CHAR`) for presence and [GAP](`GAP`)
```
I think that for consistency it should be `crate::alphabets::GAP`, same as for `AMB_CHAR`

junniest commented on phylo/data/sequences_DNA1_missmatched_id.fasta:
```diff
@@ -0,0 +1,8 @@
+>A
```
nitpick: silly comment, but missmatched --> mismatched

MattesMrzik replied:
thanks, I guess my spellchecker does not check file names

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -30,25 +41,134 @@ impl PairwiseAlignment {
+impl Display for MSA {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.seqs)
+        let mut aligned_records = Vec::with_capacity(self.seqs.len());
```
I think this closes a issue, doesn't it?

MattesMrzik replied:
Do you mean #62 ? yes it does.

junniest commented on phylo/src/evolutionary_models/mod.rs:
```diff
@@ -14,6 +14,7 @@ pub enum FrequencyOptimisation {
 
 pub trait EvoModel: Display + DynClone {
     fn p(&self, time: f64) -> SubstMatrix;
+    // same here
```
hm?

MattesMrzik replied:
Jep, I forgot to delete that

junniest reviewed (approved):
Looks great! Well done!

MattesMrzik committed:
added comment to default impls of trait

MattesMrzik committed:
adjusted file path in code after rename

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -107,172 +291,224 @@ impl Alignment {
+        for (node_idx, seq_map) in both_maps {
+            let id = &self.idx_to_id[usize::from(node_idx)];
+            let record = match node_idx {
+                Int(_) => self.ancestral_seqs.record_by_id(id),
```
see #97
